### ISSUE TITLE
reactor not firing after upgraded to rc2, 2015.2

### ISSUE THREAD

**Issue by mimianddaniel** [saltstack/salt#21651]
Events received at the master and salt master is able to rendered and return YAML result, but i dont see anything after wards. 

Normally, i would see salt. util event firing off an even to the target minions.

**Comment by rallytime** [comment]
@mimianddaniel Thanks for the report. I have been looking at this very issue today and am probably seeing the same thing as I have been running into trouble.

Also, are you just running off of the HEAD of 2015.2? We don't have an RC2 out yet on 2015.2: https://github.com/saltstack/salt/tags

**Comment by rallytime** [comment]
Ok, after taking another look, I am seeing this as well for certain (I thought before I might have been configuring something incorrectly). I can see the reactor firing on the master, and I can see the minion receive the event, but the actual event doesn't fire:

Here's the reactor sls on the master log:
```
[DEBUG   ] Rendered data from file: /srv/reactor/start.sls:
clean_tmp:
  local.cmd.run:
    - tgt: '*'
    - arg:
      - rm -rf /tmp/*

[DEBUG   ] Results of YAML rendering:
OrderedDict([('clean_tmp', OrderedDict([('local.cmd.run', [OrderedDict([('tgt', '*')]), OrderedDict([('arg', ['rm -rf /tmp/*'])])])]))])
```
Here's the minion:
```
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150313215611243219', 'tgt': '*', 'ret': '', 'user': 'root', 'arg': ['rm -rf /tmp/*'], 'fun': 'cmd.run'}
[INFO    ] Starting a new job with PID 30444
[INFO    ] Executing command 'rm -rf /tmp/*' in directory '/root'
[DEBUG   ] output:
[INFO    ] Returning information for job: 20150313215611243219
```
But then there are still files in `/tmp/*`. (My simple test here is supposed to clean out the tmp dir on the minion when it starts up.

**Comment by mimianddaniel** [comment]
Just running head of 201502.
Oddly, for me i dont see the event logged in the minion log at all. I will create similar reactor and try and see what results i get on my side.

**Comment by cachedout** [comment]
I cannot verify this. The reactor works fine for me. I see events handled and processed just as they have been.

**Comment by rallytime** [comment]
On second thought, what I am seeing is not the same thing that was originally reported. I will look at this some more and try to see what I am doing incorrectly.

@mimianddaniel Can you give some examples of what your are seeing (or not seeing) and what you're expecting to see/happen?

**Comment by mimianddaniel** [comment]
reactor that used to work on 201472 is not working after upgrading to 201502. 

on 201472, i can see the events received causes properly YAML being rendered and triggering events to the minions and the event shows up on the minion logs as well.

On 20152, I can see the events received and proper YAML being rendered but nothing afterward, neither on the master log or minion's log. 

I have tried creating different reactor and same thing.

**Comment by mimianddaniel** [comment]
Salt: 2015.2.0
         Python: 2.7.9 (default, Dec 30 2014, 13:07:54)
         Jinja2: 2.7.3
       M2Crypto: 0.22
 msgpack-python: 0.4.2
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.4.1
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed

**Comment by cachedout** [comment]
@mimianddaniel What do you mean by "proper YAML being rendered"? Are you saying that when a matching event is received that a reactor *does* evaluate the desginated block but that no job is fired? It would probably help if we could see a simplified test case and/or some master debug logs for the period wherein the event is received. Thanks!

**Comment by rallytime** [comment]
@cachedout I am still not able to get the reactor to execute on the 2015.2 branch. I checked over all of my configs to make sure I hadn't done something wrong, but could not find anything. I then tested the exact same configuration as before, but on an earlier release. Everything is working as expected on 2014.7.2 as well as the HEAD of the 2014.7 branch.

**Comment by mimianddaniel** [comment]
2015-03-13 22:22:36,798 [salt.loaded.int.render.yaml                ][DEBUG
  ] Results of YAML rendering:

OrderedDict([('update_salt', OrderedDict([('local.cmd.run',
[OrderedDict([('tgt', '.*(kvmctl|saltmaster).*')]),
OrderedDict([('expr_form', 'pcre')]), OrderedDict([('arg',
['/usr/local/bin/update-salt.sh'])])])]))])


After this, i would see the following in the 201472,

AES payload received with command _minion_event

2015-03-13 22:22:36,778 [salt.utils.event                           ][DEBUG
  ] Sending event - data = etc ...

But i do not see this on the 20152, nor do i see any events from the target
minions listed in the YAML rendering.  I see that Nicole can see the events
being received on the minion but i dont even see that...

**Comment by rallytime** [comment]
I think I might be hitting a separate issue, which I have filed in #21698. 

@mimianddaniel Before we can move forward with debugging your issue, we'll need a sample of the reactor file you're using that can give us a way to try and reproduce this, since neither I nor @cachedout are able to reproduce this. My reactor example is based on minion start-up...perhaps you are triggering the reactor with another kind of event? That information would also be helpful in reproducing your error.

Also, are you *specifically* looking for the "AES payload received" text in the minion debug logs? I do not see that specific text, and this is likely due to some pretty large refactors we have had around the loader and scheduler, etc. However, I do see the following information on the master and minion logs:
master:
```
[DEBUG   ] Sending event - data = {'tgt_type': 'glob', 'jid': '20150316185335740225', 'tgt': '*', '_stamp': '2015-03-16T18:53:35.740829', 'user': 'root', 'arg': ['rm -rf /tmp/foo'], 'fun': 'cmd.run', 'minions': ['nt-master']}
[INFO    ] User root Published command cmd.run with jid 20150316185335740225
[DEBUG   ] Published command details {'tgt_type': 'glob', 'jid': '20150316185335740225', 'tgt': '*', 'ret': '', 'user': 'root', 'arg': ['rm -rf /tmp/foo'], 'fun': 'cmd.run'}
[DEBUG   ] Gathering reactors for tag salt/event/new_client
[DEBUG   ] Gathering reactors for tag 20150316185335740225
[DEBUG   ] Gathering reactors for tag salt/job/20150316185335740225/new
[INFO    ] AES payload received with command _return
[DEBUG   ] LazyLoaded local_cache.prep_jid
[INFO    ] Got return from nt-master for job 20150316185335740225
[DEBUG   ] Sending event - data = {'fun_args': ['rm -rf /tmp/foo'], 'jid': '20150316185335740225', 'return': '', 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2015-03-16T18:53:35.801881', 'fun': 'cmd.run', 'id': 'nt-master'}
[DEBUG   ] Gathering reactors for tag salt/job/20150316185335740225/ret/nt-master
[INFO    ] Clear payload received with command _auth
```
minion:
```
[INFO    ] User root Executing command cmd.run with jid 20150316184206483633
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150316184206483633', 'tgt': '*', 'ret': '', 'user': 'root', 'arg': ['rm -rf /tmp/foo'], 'fun': 'cmd.run'}
[INFO    ] Starting a new job with PID 10925
[INFO    ] Executing command 'rm -rf /tmp/foo' in directory '/root'
[DEBUG   ] output:
[INFO    ] Returning information for job: 20150316184206483633
```

**Comment by mimianddaniel** [comment]
git.sls

{% if data['tag'] == 'git/update_salt' %}
update_salt:
  local.cmd.run:
    - tgt: '.*(kvmctl|saltmaster).*'
    - expr_form: pcre
    - arg:
      - /usr/local/bin/update-salt.sh
{% endif %}

reactor.conf
reactor:
  - 'git/*':
    - salt://reactor/git.sls

and of course we have a custom event trigger on a minion for this. That
part works fine as we can see the events coming in on the master.
AES payload received i see is in the master log file like shown in your
master log as well "

[INFO    ] AES payload received with command _return

**Comment by mimianddaniel** [comment]
curious if there has been any progress on this

**Comment by jfindlay** [comment]
@mimianddaniel, not yet.  We're working on getting a couple of releases out and then I'll be able to spend some time on this, thanks.

**Comment by mimianddaniel** [comment]
So have an update on this issue: we run master of master setup and i notice that if point the event trigger to intermediary master, the reactor works properly but if i point it to the master of the master, it doesnt. 
Intermediary master is running 2015.2.

**Comment by mimianddaniel** [comment]
so i traced my problem to this commit : 2241444f88ceade11a75fc88a20371d2d1acdae5 in rc1 and devel. 
once i remove this commit from the reactor.py. my reactor is happy again. so the reactor we have is to update the repo on all the saltmasters.  Reading the commit msg, the intent was just that? 
Would the event_user better serve as a variable out of the code ??

**Comment by jfindlay** [comment]
ping @jacksontj

**Comment by jacksontj** [comment]
The intent of that patch I think is fairly well explained in the commit message. The problem we ran into before was if you had a reactor which reacted to '*' events. In that case the reactor would start a runner (or something similar), fire an event saying it started, then react to that event. This would continue forever-- and you would end up spinning forever. Now, its possible we are running into some permissions issue or something with that user-- but it shouldn't actually check any auth at that point in the code (as @cachedout had mentioned it working, I had checked at least the basic cases as well).

I should be able to take a look at this tomorrow morning-- hopefully its something simple ;)

**Comment by jacksontj** [comment]
Turns out this is a bit messier than we expected :/ The publish interface (cmd/local) uses a different ACL system to verify the user-- which isn't bypassed by the reactor. So, for now I've opened #21987 to scope this user addition to *only* runner/wheel jobs (which do work fine). This means you can easily have infinite recursion if you ping minions, but I also found that our new event client event is also causing issues... so infinite recursion is easy if you have a rector which reacts to '*'.

**Comment by jacksontj** [comment]
#21987 is merged, this issue should be fixed. I'll have to dig in more to figure out a secure way of allowing "reactor" user to publish jobs :/

**Comment by jfindlay** [comment]
@jacksontj, thanks for your work on this.

**Comment by cachedout** [comment]
+1. Thank you as always @jacksontj.

**Comment by cachedout** [comment]
Since we've had a fix merged for a week now, I'm going to go ahead and close this issue. Thanks, all.
